pinot-common/src/main/java/com/linkedin/pinot/common/data/FieldSpec.java

```diff
@@ -131,60 +129,37 @@ public boolean isSingleValueField() {
   public void setDefaultNullValue(Object defaultNullValue) {
     Preconditions.checkNotNull(defaultNullValue);
 
-    _isDefaultNullValueSet = true;
-
-    if (_isDataTypeSet) {
-      defaultNullValueFromString(defaultNullValue.toString());
-    } else {
-      // Set default null value as the value passed in, and might reset when setDataType() get called.
-      _defaultNullValue = defaultNullValue;
-    }
-  }
-
-  /**
-   * Helper function to convert the string format default null value to the correct data type.
-   *
-   * @param value string format of the default null value.
-   */
-  protected void defaultNullValueFromString(String value) {
-    DataType dataType = getDataType();
-    switch (dataType) {
-      case INT:
-        _defaultNullValue = Integer.valueOf(value);
-        return;
-      case LONG:
-        _defaultNullValue = Long.valueOf(value);
-        return;
-      case FLOAT:
-        _defaultNullValue = Float.valueOf(value);
-        return;
-      case DOUBLE:
-        _defaultNullValue = Double.valueOf(value);
-        return;
-      case STRING:
-      case BOOLEAN:
-        _defaultNullValue = value;
-        return;
-      default:
-        throw new UnsupportedOperationException("Unknown data type " + dataType);
-    }
-  }
-
-  /**
-   * Helper function to update default null value according to the data type.
-   */
-  protected void updateDefaultNullValue() {
-    defaultNullValueFromString(_defaultNullValue.toString());
+    _defaultNullValue = defaultNullValue.toString();
   }
 
+  // BOOLEAN, BYTE, CHAR, SHORT support is a temporary work around, will remove it after using canonical schema.
   public Object getDefaultNullValue() {
     if (_defaultNullValue != null) {
-      return _defaultNullValue;
+      switch (_dataType) {
+        case BYTE:
+        case SHORT:
+        case INT:
+          return Integer.valueOf(_defaultNullValue);
```

Comment:
Would be better to cache this value in a transient field, as this conversion will be invoked on every row where there is a null value.